test/distributed/test_c10d.py

```diff
@@ -3048,32 +3052,86 @@ def fut_then(fut):
 
-        # check whether the grads are equal to what then callback returns.
+        # check whether the grads are equal to what simple_hook's then callback returns.
         # without the comm_hook, result would be 0.25 * torch.ones(2, 2).
-        [self.assertEqual(p.grad, 2 * torch.ones(2, 2)) for p in model.parameters()]
+        self._run_and_verify_hook(gpu_model, 8 , 2 * torch.ones(2, 2))
 
-    @requires_gloo()
+    @requires_nccl()
     @skip_if_lt_x_gpu(2)
-    def test_ddp_comm_hook_future_passing_gpu_gloo(self):
+    def test_ddp_comm_hook_future_passing_gpu_nccl(self):
         """
-        This unit test executes _test_ddp_comm_hook_future_passing_gpu using gloo backend.
+        This unit test verifies whether the Future object is passed properly using nccl backend.
+        The hook callback function creates a Future object and sets a value to it.
         """
-        self._test_ddp_comm_hook_future_passing_gpu(c10d.ProcessGroupGloo)
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+
+        # Get GPU model with simple_hook registered.
+        gpu_model = self._gpu_model_with_ddp_comm_hook(process_group, self._simple_hook)
+
+        # check whether the grads are equal to what simple_hook's then callback returns.
+        # without the comm_hook, result would be 0.25 * torch.ones(2, 2).
+        self._run_and_verify_hook(gpu_model, 8 , 2 * torch.ones(2, 2))
 
     @requires_nccl()
     @skip_if_lt_x_gpu(2)
-    def test_ddp_comm_hook_future_passing_gpu_nccl(self):
+    def test_ddp_comm_hook_allreduce_hook_nccl(self):
         """
-        This unit test executes _test_ddp_comm_hook_future_passing_gpu using nccl backend.
+        This unit test verifies whether a DDP communication hook that just calls
+        allreduce gives the same result result with the case of no hook registered.
+        Without the then callback, the future_value in reducer is no longer
+        a PyObject, and this unit test verifies future_value is properly checked.
         """
-        self._test_ddp_comm_hook_future_passing_gpu(c10d.ProcessGroupNCCL)
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+
+        def allreduce_hook(state: object, bucket: dist.GradBucket) -> torch.futures.Future:
+            return process_group.allreduce(bucket.get_tensors()).get_future()
+
+        # Get GPU model with allreduce_hook registered.
+        gpu_model = self._gpu_model_with_ddp_comm_hook(process_group, allreduce_hook)
+
+        # check whether the grads are equal to what DDP without hook would return.
+        self._run_and_verify_hook(gpu_model, 8, 0.25 * torch.ones(2, 2))
+
+    @requires_nccl()
+    @skip_if_lt_x_gpu(2)
+    def test_ddp_comm_hook_allreduce_then_mult_ten_hook_nccl(self):
```

Comment:
nit: s/ten/then

Reply:
ten is actually 10, I'll just replace this as `test_ddp_comm_hook_allreduce_then_mult_hook_nccl`